common/lib/plugins/failover2/failover2_plugin.ts

```diff
@@ -0,0 +1,502 @@
+        const currentHostInfo = this.pluginService.getCurrentHostInfo();
+        if (currentHostInfo !== null) {
+          this.pluginService.setAvailability(currentHostInfo.allAliases ?? new Set(), HostAvailability.NOT_AVAILABLE);
+        }
+
+        this._lastError = e;
+        await this.pickNewConnection();
+      }
+
+      throw e;
+    }
+  }
+
+  async failover(failedHost: HostInfo) {
+    this.pluginService.setAvailability(failedHost.allAliases, HostAvailability.NOT_AVAILABLE); // checking if needed
+
+    if (this.failoverMode === FailoverMode.STRICT_WRITER) {
+      await this.failoverWriter();
+    } else {
+      await this.failoverReader();
+    }
+
+    if (this._isInTransaction || this.pluginService.isInTransaction()) {
+      // "Transaction resolution unknown. Please re-configure session state if required and try
+      // restarting transaction."
+      logger.debug(Messages.get("Failover.transactionResolutionUnknownError"));
+      throw new TransactionResolutionUnknownError(Messages.get("Failover.transactionResolutionUnknownError"));
+    } else {
+      // "The active SQL connection has changed due to a connection failure. Please re-configure
+      // session state if required."
+      throw new FailoverSuccessError(Messages.get("Failover.connectionChangedError"));
+    }
+  }
+
+  async failoverReader() {
+    const telemetryFactory = this.pluginService.getTelemetryFactory();
+    const telemetryContext = telemetryFactory.openTelemetryContext(Failover2Plugin.TELEMETRY_READER_FAILOVER, TelemetryTraceLevel.NESTED);
+    this.failoverReaderTriggeredCounter.inc();
+
+    const oldAliases = this.pluginService.getCurrentHostInfo()?.allAliases ?? new Set();
+    const failoverEndTimeMs = Date.now() + this.failoverTimeoutSettingMs;
+
+    try {
+      logger.debug(Messages.get("Failover.startReaderFailover"));
+      await telemetryContext.start(async () => {
+        if (!(await this.pluginService.initiateTopologyUpdate(false, 0))) {
+          // Unable to establish SQL connection to an instance.
+          this.failoverReaderFailedCounter.inc();
+          logger.warn(Messages.get("Failover2.unableToFetchTopology"));
+          throw new FailoverFailedError(Messages.get("Failover.unableToConnectToReader"));
+        }
+
+        // Signal to connect that this is an internal call and does not require additional processing.
+        const copyProps = new Map<string, any>(this._properties);
+        copyProps.set(Failover2Plugin.INTERNAL_CONNECT_PROPERTY_NAME, true);
+
+        while (Date.now() < failoverEndTimeMs) {
+          const hosts: HostInfo[] = this.pluginService.getHosts();
+          const hostsByPriority = this.getHostsByPriority(hosts);
+          let readerCandidateClient: ClientWrapper = null;
+          let readerCandidateHostInfo: HostInfo = null;
+
+          while (readerCandidateClient === null && hostsByPriority.length > 0 && Date.now() < failoverEndTimeMs) {
```

Comment:
```suggestion
          while (hostsByPriority.length > 0 && Date.now() < failoverEndTimeMs) {
```
`readerCandidateClient` will always be `null` when we get to this condition.